### DISCUSSION TITLE
Feedback on dev version 2.2 with GTK4

### DISCUSSION THREAD

**Discussion by iksi4prs** [PintaProject/Pinta#654] [General]
Hi,
@cameronwhite 
I encounter some things which behave in a "not friendly" way, but maybe are still under development.
Any dedicated place to give feedback on version 2,.2 with GTK4 ?
or here it's good enough ?

**Comment by cameronwhite** [comment]
There isn't really a dedicated place - here is good, or the `Issues` page if it's a more concrete bug to be fixed.
I'm hoping to do a 2.2 release fairly soon, so it would definitely be good to sort out any big issues that you're encountering!

**Reply by iksi4prs** [reply]
Hi @cameronwhite ,
Maybe this is still "wip", so I write it here.
This is related to screen resolution and screen ratio,
When screen is not "high" enough, there's no place for all buttons in toolbar of tools.
This was also in the 2.1 version, but now the behavior has changed.
In version 2.1 there was a menu in the bottom to open the access that extra buttons.
In version 2.2, there is a scrollbar, to access all buttons.

The problem is this scrollbar opens every time (expected as some buttons are not visible),
and because the toolbar is now narrower, it covers almost half of the area of buttons, so cannot click them.
(hard to see in picture, i think it captures the mouse also in areas that are not in the scrollbar that is painted).

So every time I want to click a button, the scroll button opens, and I then need to move the mouse back to right, and then slowly-slowly-gently move it to the left to the toolbar, without "awakening" the scrollbar.

See attached image.
I think a partial solution will be that toolbar will be wider as before. 
The scrollbar will still open every time (which is annoying), but at least the buttons will be clickable.
If possible to put the buttons in 2 columns that would better in my opinion, but I guess some people that are already used it wont like this idea.

![image](https://github.com/PintaProject/Pinta/assets/65832579/165f5d12-5a40-4427-a58d-917632e42ea8)

**Reply by cameronwhite** [reply]
Thanks!
I've set up a branch with some experimental changes - https://github.com/PintaProject/Pinta/actions/runs/7405489857 - let me know what you think

It'll end up looking something like this, which I think generally works okay, but can look a bit weird if there are only a couple extra buttons in the next column since GTK's layout doesn't evenly distribute between the columns :(

<img width="341" alt="layout" src="https://github.com/PintaProject/Pinta/assets/303951/f45bcd8a-3937-431f-8637-7cae069c6fe9">

**Reply by iksi4prs** [reply]
@cameronwhite , looks good, but I didn't have much time to check.
After few seconds the app crashes, and fails to start again.
I found these events on Windows's events viewer.
If it doesn't help you, I can try build on my machine (I **don't** have .net 8 installed, hence didn't try so far)

```
Application: Pinta.exe
CoreCLR Version: 8.0.23.53103
.NET Version: 8.0.0
Description: The process was terminated due to an unhandled exception.
Exception Info: System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
Stack:
   at GObject.Internal.Object.RemoveToggleRef(IntPtr, GObject.Internal.ToggleNotify, IntPtr)
   at GObject.Internal.Object.RemoveToggleRef(IntPtr, GObject.Internal.ToggleNotify, IntPtr)
   at GObject.Internal.ObjectMapper.Unmap(IntPtr)
   at GObject.Internal.ObjectHandle.ReleaseHandle()
   at System.Runtime.InteropServices.SafeHandle.InternalRelease(Boolean)
   at System.Runtime.InteropServices.SafeHandle.Finalize()
```

and also
```
Faulting application name: Pinta.exe, version: 2.2.0.0, time stamp: 0x65410000
Faulting module name: coreclr.dll, version: 8.0.23.53103, time stamp: 0x65411660
Exception code: 0xc0000005
Fault offset: 0x00000000001e337a
Faulting process id: 0x0xBC80
Faulting application start time: 0x0x1DA3F1C8D463DE9
Faulting application path: C:\Program Files\Pinta\bin\Pinta.exe
Faulting module path: C:\Program Files\Pinta\bin\coreclr.dll
Report Id: fa0399c5-aff9-4e9b-b43c-0b1429429ed9
Faulting package full name: 
Faulting package-relative application ID: 
```

and 
```
Faulting application name: Pinta.exe, version: 2.1.1.0, time stamp: 0x63c9df9f
Faulting module name: KERNELBASE.dll, version: 10.0.22621.2792, time stamp: 0x3091b6fb
Exception code: 0x80000003
Fault offset: 0x000000000010cd82
Faulting process id: 0x0x6A54
Faulting application start time: 0x0x1DA3F1CC7555BDF
Faulting application path: C:\Program Files\Pinta\Pinta.exe
Faulting module path: C:\WINDOWS\System32\KERNELBASE.dll
Report Id: 8d3cb891-26e0-42d4-943e-630d861f029b
Faulting package full name: 
Faulting package-relative application ID: 
```

**Reply by badcel** [reply]
Oh this could be a bug in @GirCore. I just had a similar behavior on windows only while implementing the record improvements.

It happened if code was allocated via the dotnet `Marshal` class but was attempted to be freed via a native glib function.

**Reply by cameronwhite** [reply]
> Oh this could be a bug in @gircore. I just had a similar behavior on windows only while implementing the record improvements.
> 
> It happened if code was allocated via the dotnet `Marshal` class but was attempted to be freed via a native glib function.

We also have some manual bindings in Pinta, so there could also be a bug with one of those

**Reply by iksi4prs** [reply]
@cameronwhite , should I open an issue for the "unhandled exception", or you want to wait for final build of 2.2 ?

**Reply by cameronwhite** [reply]
Yes, please do file an issue - this is definitely something I'd like to fix before the release. I haven't had much luck yet reproducing it on my Windows machine though :/

**Reply by iksi4prs** [reply]
Created issue:
https://github.com/PintaProject/Pinta/issues/674

**Reply by iksi4prs** [reply]
@cameronwhite ,
This is related to behavior's of the 2 columns toolbar.
When working (not related to crash), I've noticed that depending on the selected tool,
some ratios of border/margins/padding change.
If you changes tools quickly it looks like the image is "jumping".
See attached image.
You can see in points A and B, that the sizes are different (need to zoom, its only 2-3 pixels) based on tool used.
Maybe this is a result of the height of the "Replace" combox-box ? it looks higher than others.

![image](https://github.com/PintaProject/Pinta/assets/65832579/2513d126-8e8b-481a-9a45-392e5368870f)

**Reply by cameronwhite** [reply]
I think this might have been fixed already in the master branch with 3e3d3f32, which added a slightly larger default height so that it doesn't jump when the larger combo boxes appear
